[ONNX] Add full checker mode in torch.onnx.export

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84789
* __->__ #83186

Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel @ezyang @gchanan

titaiwangms committed:
add full checker mode

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.nn.export"


Why:
1. Currently, the full_check mode in `_C._check_onnx_proto` does nothing, as `onnx::shape_inference::InferShapes(model)` is included in `onnx::checker::check_model(model)`
2. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
3. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.nn.export"


Why:
1. Currently, the full_check mode in `_C._check_onnx_proto` does nothing, as `onnx::shape_inference::InferShapes(model)` is included in `onnx::checker::check_model(model)`
2. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
3. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.

CI should pass after #83201 

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.nn.export"


Why:
1. Currently, the full_check mode in `_C._check_onnx_proto` does nothing, as `onnx::shape_inference::InferShapes(model)` is included in `onnx::checker::check_model(model)`
2. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
3. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.

CI should pass after #83201 

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.nn.export"


Why:
1. Currently, the full_check mode in `_C._check_onnx_proto` does nothing, as `onnx::shape_inference::InferShapes(model)` is included in `onnx::checker::check_model(model)`
2. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
3. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.nn.export"


Why:
1. Currently, the full_check mode in `_C._check_onnx_proto` does nothing, as `onnx::shape_inference::InferShapes(model)` is included in `onnx::checker::check_model(model)`
2. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
3. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.nn.export"


Why:
1. Currently, the full_check mode in `_C._check_onnx_proto` does nothing, as `onnx::shape_inference::InferShapes(model)` is included in `onnx::checker::check_model(model)`
2. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
3. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.nn.export"


Why:
1. Currently, the full_check mode in `_C._check_onnx_proto` does nothing, as `onnx::shape_inference::InferShapes(model)` is included in `onnx::checker::check_model(model)`
2. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
3. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.nn.export"


Why:
1. Currently, the full_check mode in `_C._check_onnx_proto` does nothing, as `onnx::shape_inference::InferShapes(model)` is included in `onnx::checker::check_model(model)`
2. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
3. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.nn.export"


Why:
1. Currently, the full_check mode in `_C._check_onnx_proto` does nothing, as `onnx::shape_inference::InferShapes(model)` is included in `onnx::checker::check_model(model)`
2. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
3. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.


[ghstack-poisoned]

thiagocrepaldi commented on torch/csrc/jit/serialization/export.cpp:
```diff
@@ -1366,11 +1366,8 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
-  onnx::checker::check_model(model);
-
-  if (full_check) {
-    onnx::shape_inference::InferShapes(model);
```
why was this removed?

AllenTiTaiWang replied:
onnx::shape_inference::InferShapes(model) is included in onnx::checker::check_model(model) in the latest version.
https://github.com/onnx/onnx/blob/f0d824c0ead2825f7c24790c05f985ad1fb909f2/onnx/checker.cc#L941-L958

AllenTiTaiWang replied:
I thought check_model includes regular InferShapes, but it's not. I will later change it to:
```python
onnx::checker::check_model(model, full_check);
if (!full_check) {
  onnx::shape_inference::InferShapes(model);
}
```

AllenTiTaiWang replied:
Done

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -470,7 +471,10 @@ def forward(self, x):
-
+        full_model_check (bool, default False): if True, ONNX checker will give the exported model a
+            strict shape type inference check to make sure it's a valid ONNX model. Notice that only a
+            valid ONNX model can be executed on ONNXRUNTIME. Default to False, as users might run the
```
I wonder how it plays with `operator_export_type` that are not `ONNX`?

AllenTiTaiWang replied:
It won't check non-ONNX according to:
https://github.com/pytorch/pytorch/blob/6865a35c08d9e1ad7fd4daf481fa2803a8655f00/torch/onnx/utils.py#L1543-L1549

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -200,6 +200,7 @@ def export(
     keep_initializers_as_inputs: Optional[bool] = None,
     custom_opsets: Optional[Mapping[str, int]] = None,
     export_modules_as_functions: Union[bool, Collection[Type[torch.nn.Module]]] = False,
+    full_model_check: bool = False,
```
We'd try avoid adding new arguments in an effort to not make the export api more complicated than it already is. Also a couple of alternatives to consider:

* Always run checker but have it emit warnings instead of errors. Removing the need of this argument.
* Always run checker but have it emit warnings instead of errors based on a new flag. With similar effect like "strict mode", this flag allows suppression of certain errors.

AllenTiTaiWang replied:
Done

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.


[ghstack-poisoned]

AllenTiTaiWang commented:
cc @BowenBao @shubhambhokare1

justinchuby commented on torch/csrc/jit/serialization/export.cpp:
```diff
@@ -1366,10 +1366,22 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
+      onnx::checker::check_model(model, full_check);
+    } catch (const onnx::InferenceError& ex) {
+      TORCH_WARN(
+          "This is an invalid ONNX model with failed ONNX shape type inference. It is not executable by ONNXRUNTIME platform. More detail: ",
```
```suggestion
          "The exported ONNX model failed ONNX shape inference. The model will not be executable by the ONNX Runtime. If this is unintended and you believe there is a bug, please report an issue at https://github.com/pytorch/pytorch/issues. Error reported by the ONNX checker: ",
```

suggested wording (be sure to break the lines if too long)

AllenTiTaiWang replied:
Done

justinchuby commented on torch/csrc/jit/serialization/export.cpp:
```diff
@@ -1366,10 +1366,22 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
-    onnx::shape_inference::InferShapes(model);
+    try {
+      onnx::checker::check_model(model, full_check);
+    } catch (const onnx::InferenceError& ex) {
```
Nice warning w/ original message

justinchuby commented on torch/csrc/onnx/init.cpp:
```diff
@@ -229,9 +229,10 @@ void initONNXBindings(PyObject* module) {
-      [](const std::string& proto_string, bool full_check) {
-        check_onnx_proto(proto_string, full_check);
-      },
+      ::torch::wrap_pybind_function(
+          [](const std::string& proto_string, bool full_check) {
+            check_onnx_proto(proto_string, full_check);
+          }),
```
Do we need a lambda here? Can this be `::torch::wrap_pybind_function(check_onnx_proto)`?

Just wondering, since there may be pybind limitations I am not aware of

AllenTiTaiWang replied:
Tested, and it made no difference, so I modified it as suggested.

AllenTiTaiWang replied:
Sorry, this is needed, since it passes parameter names to pybind and let users consume them in python. I think we should keep it.

justinchuby replied:
Can you share a link if there's a doc?

thiagocrepaldi replied:
> Do we need a lambda here? Can this be `::torch::wrap_pybind_function(check_onnx_proto)`?
> 
> Just wondering, since there may be pybind limitations I am not aware of

@justinchuby Do you know what `::torch::wrap_pybind_function(check_onnx_proto)`? Adding code we dont know what it does is dangerous. The safer bet is, if it is working and you dont know what the changes would cause, better not to touch it :)

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -470,7 +470,6 @@ def forward(self, x):
             * ``True``: export all ``nn.Module`` forward calls as local function nodes.
             * Set of type of nn.Module: export ``nn.Module`` forward calls as local function nodes,
               only if the type of the ``nn.Module`` is found in the set.
-
```
I would put this line break back

AllenTiTaiWang replied:
Done

justinchuby commented on torch/csrc/jit/serialization/export.cpp:
```diff
@@ -1366,10 +1366,22 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
-
+  onnx::shape_inference::InferShapes(model);
+  // full check including strict shape inference check
+  // Notice that onnx::checker::check_model doesn't
```
```suggestion
  // NOTE: onnx::checker::check_model doesn't
```

justinchuby replied:
I would reword this to make the intention clear

AllenTiTaiWang replied:
Done

justinchuby commented on torch/csrc/jit/serialization/export.cpp:
```diff
@@ -1366,10 +1366,22 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
+  // include non-strict model onnx::shape_inference::InferShapes(model).
+  // That's why we need to write another onnx::checker::check_model(model,
+  // full_check) here.
   if (full_check) {
```
If full_check is always true, should we remove it from the function?

AllenTiTaiWang replied:
Leaving this for now to see what others think.

thiagocrepaldi replied:
Why not use `torch.onnx._globals.py::_InternalGlobals.onnx_shape_inference` or similar to control this behavior?

Shape inference is not required by ONNX spec, it is an optional feature. If an invalid ONNX graph is being generated in ONNX converter, we should fix it, regardless of running or not inference shape every time

justinchuby reviewed (changes requested):
changes pending

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.


[ghstack-poisoned]

justinchuby commented:
Optional means a field can take None as an argument. If a paean is defaulted to True, we should annotate as `bool`, since None as an input is usually not meaningful

justinchuby commented on torch/csrc/jit/serialization/export.cpp:
```diff
@@ -1370,10 +1370,26 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
+  // Note: onnx::checker::check_model doesn't
+  // include non-strict model onnx::shape_inference::InferShapes(model).
+  // That's why we need to write another onnx::checker::check_model(model,
+  // full_check) here.
```
Just checking - was this reworded?

AllenTiTaiWang replied:
I think it's Notice that -> Note:

AllenTiTaiWang replied:
reword: 
1. clarify non-strict mode
2. call check_model twice reason

AllenTiTaiWang commented:
> Optional means a field can take None as an argument. If a boolean is defaulted to True, we usually annotate it as `bool`, since None as an input is usually not meaningful

Is this comment actually for another PR?

thiagocrepaldi commented on torch/csrc/onnx/init.cpp:
```diff
@@ -235,9 +235,10 @@ void initONNXBindings(PyObject* module) {
-      [](const std::string& proto_string, bool full_check) {
-        check_onnx_proto(proto_string, full_check);
-      },
+      ::torch::wrap_pybind_function(
```
Is this really needed? Many bindings in this file do not use. Try to understand what `::torch::wrap_pybind_function` does before start using it. If it is not needed, the lesser code, the better.

justinchuby replied:
Yes. See https://github.com/pytorch/pytorch/pull/78441

AllenTiTaiWang replied:
`::torch::wrap_pybind_function` helps to wrap the error into warning which help the PR achieve the goal of "not error out the exporter, but give the warning of why this model doesn't pass `strict_mode` shape type inference"

thiagocrepaldi reviewed (changes requested):
Just would like to make sure the offending operator is being fixed (in this PR or another) in addition to force a check that can break other backends

thiagocrepaldi commented on torch/csrc/jit/serialization/export.cpp:
```diff
@@ -1370,10 +1370,26 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
+  // baseline check
   onnx::checker::check_model(model);
-
+  onnx::shape_inference::InferShapes(model);
```
> The thing is `stric_mode` is considered a baseline check now. If a model not passing `strict_mode` shape type inference check, it's not a valid ONNX graph.

That is my point. Why it is the baseline now? Why? It seems there is a bug in the conversion that is not being handled and we are waiting for shape inference to fix it for us.

We must export things correctly without dependencies on other APIs to fix a malformed operator. Maybe we need an actual example to see the error we are discussing

AllenTiTaiWang replied:
Resolved in offline meeting. This PR simply adds warning message, nit changing the original behavior of `check_onnx_proto`

jcwchen commented on torch/csrc/jit/serialization/export.cpp:
```diff
@@ -1370,10 +1370,26 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
   if (full_check) {
-    onnx::shape_inference::InferShapes(model);
+    try {
+      onnx::checker::check_model(model, full_check);
```
I think we can directly run shape inference with strict mode:
```suggestion
      auto* schema_registry = onnx::OpSchemaRegistry::Instance();
      onnx::ShapeInferenceOptions options{
          /*check_type=*/True,
          /*error_mode=*/True};
      onnx::shape_inference::InferShapes(
          *model_proto, schema_registry, options);
```
Otherwise it runs onnx.checker twice.

AllenTiTaiWang replied:
Nice catch

jcwchen commented on torch/csrc/jit/serialization/export.cpp:
```diff
@@ -1370,10 +1370,26 @@ void check_onnx_proto(const std::string& proto_string, bool full_check) {
+          "The model will not be executable by the ONNX Runtime."
+          "If this is unintended and you believe there is a bug,"
+          "please report an issue at https://github.com/pytorch/pytorch/issues."
+          "Error reported by the ONNX checker: ",
```
More precise:
```suggestion
          "Error reported by the strict ONNX shape inference: ",
```

AllenTiTaiWang replied:
Done

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

AllenTiTaiWang commented:
> Just would like to make sure the offending operator is being fixed (in this PR or another) in addition to force a check that can break other backends

This warning message will be helpful for this case https://github.com/pytorch/pytorch/issues/83600 (the fix https://github.com/pytorch/pytorch/pull/83599)
With the warning message, we can identify the failed shape type inference ahead of onnxruntime.

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

wschin commented:
Maybe let's use exporting mode to add a new mode to support the new evil mode?

AllenTiTaiWang commented:
> Maybe let's use exporting mode to add a new mode to support the new evil mode?

To avoid the confusion, I will have another PR doing this. This further work can be track by #84884

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

thiagocrepaldi commented:
Is there anything blocking this PR? AFAIK, the final version should 1) keep the current behavior and 2) print a useful warning for ONNX Runtime users, allowing them to identify the runtime issue.

Is that it?

AllenTiTaiWang commented:
This PR needs ONNX version bump.

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Depends: https://github.com/pytorch/pytorch/pull/83201
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Depends: https://github.com/pytorch/pytorch/pull/83201
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Depends: https://github.com/pytorch/pytorch/pull/83201
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Depends: https://github.com/pytorch/pytorch/pull/83201
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Depends: https://github.com/pytorch/pytorch/pull/83201
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Depends: https://github.com/pytorch/pytorch/pull/83201
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Depends: https://github.com/pytorch/pytorch/pull/83201
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Depends: https://github.com/pytorch/pytorch/pull/83201
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add full checker mode in torch.onnx.export"


Fix #82589
Why:
1. **full_check** works in `onnx::checker::check_model` function as it turns on **strict_mode** in `onnx::shape_inference::InferShapes()` which I think that was the intention of this part of code.
2. **strict_mode** catches failed shape type inference (invalid ONNX model from onnx perspective) and ONNXRUNTIME can't run these invalid models, as ONNXRUNTIME actually rely on ONNX shape type inference to optimize ONNX graph. Why we don't set it True for default? >>> some of existing users use other platform, such as caffe2 to run ONNX model which doesn't need valid ONNX model to run.
3. This PR doesn't change the original behavior of `check_onnx_proto`, but add a warning message for those models which can't pass strict shape type inference, saying the models would fail on onnxruntime.

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge -g

titaiwangms commented:
> ## Merge failed
> **Reason**: GraphQL query fragment PRCheckSuites on CheckSuiteConnection { edges { node { app { name databaseId } workflowRun { workflow { name } url } checkRuns(first: 50) { nodes { name conclusion detailsUrl } pageInfo { endCursor hasNextPage } } conclusion } cursor } pageInfo { hasNextPage } }
> 
> query ($owner: String!, $name: String!, $number: Int!, $cursor: String!) { repository(name: $name, owner: $owner) { pullRequest(number: $number) { commits(last: 1) { nodes { commit { oid checkSuites(first: 10, after: $cursor) { ...PRCheckSuites } } } } } } } , args {'name': 'pytorch', 'owner': 'pytorch', 'number': 83186, 'cursor': 'Y3Vyc29yOnYyOpHPAAAAAoddcug='} failed: [{'message': 'Something went wrong while executing your query. Please include `07C3:7CCA:43256AB:88A6B31:63E4169C` when reporting this issue.'}]
> 
> Details for Dev Infra team

Hi @kit1980,

Do you know what issue this might be?

kit1980 commented:
Looks like GitHub hiccup?
Let's try again.

kit1980 commented:
@pytorchbot merge

kit1980 commented on torch/csrc/jit/serialization/export.cpp:
```diff
@@ -1380,16 +1380,36 @@ std::string serialize_model_proto_to_string(
 }
 
-void check_onnx_proto(const std::string& proto_string, bool full_check) {
+void check_onnx_proto(const std::string& proto_string) {
```
This changes public API, no?

titaiwangms replied:
I think it does.  It's one of TORCH API. Is there anything else I should do?

kit1980 replied:
You need to follow https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy
It basically says first emit a warning for 2 releases, then can delete.
But for this function specifically I'm not sure if it's even used by anyone.
We may need to revert if there are complains.

titaiwangms replied:
Thanks!

kit1980 replied:
I've added bc breaking labels.
